### PR TITLE
OpenACC ParallelReduce with MDRange

### PR DESCRIPTION
**Pull request by seyonglee**
Initial implementation of the OpenACC parallel_reduce() with MDRange
    - Updated the OpenACC parallel_reduce() with Range to be compatible with recent changes

### CODE REVIEW

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp [hunk 1]
```diff
@@ -0,0 +1,1590 @@
...
+
+// primary template: catch-all non-implemented custom reducers
+template <class Functor, class Reducer, class Policy,
+          bool = std::is_arithmetic_v<typename Reducer::value_type>>
+struct OpenACCParallelReduceMDRangeHelper {
+  OpenACCParallelReduceMDRangeHelper(Functor const&, Reducer const&,
+                                     Policy const&) {
+    static_assert(
+        !Kokkos::Impl::always_true<Functor>::value,
+        "Kokkos Error: not implemented reducer type for the OpenACC backend");
```
Comment (masterleinad) [inline core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp:71]:
```suggestion
        "Kokkos Error: unimplemented reducer type for the OpenACC backend");
```

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp [hunk 2]
```diff
@@ -0,0 +1,1590 @@
...
+                 ReducerType const& reducer)
+      : m_functor(functor),
+        m_policy(policy),
+        m_reducer(reducer),
+        m_result_ptr(reducer.view().data()) {}
+
+  template <class ViewType>
+  ParallelReduce(
+      const Functor& functor, const Policy& policy, const ViewType& result,
+      std::enable_if_t<Kokkos::is_view<ViewType>::value, void*> = nullptr)
```
Comment (masterleinad) [inline core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp:104]:
```suggestion
      std::enable_if_t<Kokkos::is_view_v<ViewType>>* = nullptr)
```

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp [hunk 3]
```diff
@@ -0,0 +1,1590 @@
...
+      : m_functor(functor),
+        m_policy(policy),
+        m_reducer(InvalidType()),
+        m_result_ptr(result.data()) {}
+
+  void execute() {
+    static_assert(Policy::rank < 7 && Policy::rank > 1,
+                  "OpenACC Backend MDRangePolicy Error: Unsupported rank...");
```
Comment (masterleinad) [inline core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp:112]: Shouldn't this check be in the `MDRangePolicy` class? Is this restriction specific to the `OpenACC` backend?

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp [hunk 4]
```diff
@@ -0,0 +1,1590 @@
...
+    typename Analysis::Reducer final_reducer(&m_functor);
+    final_reducer.init(&val);
+
+    Kokkos::Experimental::Impl::OpenACCParallelReduceMDRangeHelper(
+        Kokkos::Experimental::Impl::FunctorAdapter<Functor, Policy>(m_functor),
+        std::conditional_t<is_reducer_v<ReducerType>, ReducerType,
+                           Sum<ValueType>>(val),
+        m_policy);
+
+    *m_result_ptr = val;
```
Comment (masterleinad) [inline core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp:123]: Are you missing to call
```
final_reducer.final(&val);
```
here? Ideally, `init` and `final` should run on the device.

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp [hunk 5]
```diff
@@ -0,0 +1,1590 @@
...
+namespace Kokkos::Experimental::Impl {
+
+template <class ValueType, class Functor>
+void OpenACCParallelReduceSum(OpenACCMDRangeBegin<2> const& begin,
+                              OpenACCMDRangeEnd<2> const& end, ValueType& val,
+                              Functor const& functor, int const async_arg) {
+  auto const begin1 = begin[1];
+  auto const end1   = end[1];
+  auto const begin0 = begin[0];
+  auto const end0   = end[0];
```
Comment (masterleinad) [inline core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp:136]: Is there any benefit in using these aliases (here and below)?

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp [hunk 6]
```diff
@@ -0,0 +1,1590 @@
...
+      }
+    }
+  }
+  val = l_val;
+}
+
+template <class ValueType, class Functor>
+void OpenACCParallelReduceProd(OpenACCMDRangeBegin<2> const& begin,
+                               OpenACCMDRangeEnd<2> const& end, ValueType& val,
+                               Functor const& functor, int const async_arg) {
```
Comment (masterleinad) [inline core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp:310]: This only differs minimally from `OpenACCParallelReduceSum`. Can't you unify the definition for different kinds of reductions using a macro?

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp [hunk 7]
```diff
@@ -70,10 +70,8 @@ class Kokkos::Impl::ParallelReduce<Functor, Kokkos::RangePolicy<Traits...>,
                                    ReducerType, Kokkos::Experimental::OpenACC> {
   using Policy = RangePolicy<Traits...>;
 
-  using Analysis = FunctorAnalysis<
-      FunctorPatternInterface::REDUCE, Policy,
-      std::conditional_t<std::is_same_v<InvalidType, ReducerType>, Functor,
-                         ReducerType>>;
+  using Analysis =
+      FunctorAnalysis<FunctorPatternInterface::REDUCE, Policy, Functor>;
```
Comment (masterleinad) [inline core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp:74]: Why don't you take the `ReducerType` into account here anymore?

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp [hunk 8]
```diff
@@ -0,0 +1,1590 @@
...
+
+}  // namespace Kokkos::Experimental::Impl
+
+template <class Functor, class ReducerType, class... Traits>
+class Kokkos::Impl::ParallelReduce<Functor, Kokkos::MDRangePolicy<Traits...>,
+                                   ReducerType, Kokkos::Experimental::OpenACC> {
+  using Policy = MDRangePolicy<Traits...>;
+
+  using Analysis =
+      FunctorAnalysis<FunctorPatternInterface::REDUCE, Policy, Functor>;
```
Comment (masterleinad) [inline core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp:83]: Why don't you take the `ReducerType` into account here?

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp [hunk 9]
```diff
@@ -126,12 +124,13 @@ namespace Kokkos::Experimental::Impl {
...
 template <class IndexType, class ValueType, class Functor>
 void OpenACCParallelReduceSum(Schedule<Static>, int chunk_size, IndexType begin,
-                              IndexType end, ValueType& val,
+                              IndexType end, ValueType& aval,
                               Functor const& afunctor, int async_arg) {
   // FIXME_OPENACC FIXME_NVHPC workaround compiler bug (incorrect scope
   // analysis)
   // NVC++-S-1067-Cannot determine bounds for array - functor
   auto const functor(afunctor);
+  auto val = aval;
```
Comment (masterleinad) [inline core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp:133]: Why did you change this?

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp (hunk 3 above)
Comment (seyonglee) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp:112]: This is just for double-checking for the OpenACC backend. 
Whenever rank restriction in MDRangePolicy is changed, the OpenACC backend implementation should be changed too. This static_assert can detect unintended mismatch between rank restrictions in MDRangePolicy and the OpenACC backend implementation.

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp (hunk 5 above)
Comment (seyonglee) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp:136]: Local scalar variables are much easier for an OpenACC compiler to analyze their sharing rules and optimize.

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp (hunk 9 above)
Comment (seyonglee) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp:133]: Local scalar variables are much easier for an OpenACC compiler to analyze their sharing rules and optimize.

File: core/src/decl/Kokkos_Declare_OPENACC.hpp [hunk 10]
```diff
@@ -55,6 +55,7 @@
 #include <OpenACC/Kokkos_OpenACC_MDRangePolicy.hpp>
 #include <OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp>
 #include <OpenACC/Kokkos_OpenACC_ParallelFor_Team.hpp>
+#include <OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp>
```
Comment (dalg24) [inline core/src/decl/Kokkos_Declare_OPENACC.hpp:58]:
```suggestion
#include <OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp>
#include <OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp>
#include <OpenACC/Kokkos_OpenACC_ParallelFor_Team.hpp>
```

File: core/src/OpenACC/Kokkos_OpenACC_Team.hpp [hunk 11]
```diff
@@ -109,7 +109,7 @@ class OpenACCTeamMember {
   // FIXME_OPENACC: team_broadcast() is not implemented.
   template <class ValueType>
   KOKKOS_FUNCTION void team_broadcast(ValueType& value, int thread_id) const {
-    static_assert(Kokkos::Impl::always_false<ValueType>::value,
+    static_assert(!Kokkos::Impl::always_true<ValueType>::value,
```
Comment (dalg24) [inline core/src/OpenACC/Kokkos_OpenACC_Team.hpp:112]: Does not belong to this PR

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp (hunk 9 above)
Comment (dalg24) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp:133]: Please open another PR for these changes

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp (hunk 7 above)
Comment (dalg24) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp:74]: Ping

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp (hunk 8 above)
Comment (dalg24) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp:83]: PIng

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp [hunk 12]
```diff
@@ -0,0 +1,1590 @@
...
+#include <OpenACC/Kokkos_OpenACC.hpp>
+#include <OpenACC/Kokkos_OpenACC_FunctorAdapter.hpp>
+#include <OpenACC/Kokkos_OpenACC_MDRangePolicy.hpp>
+#include <Kokkos_Parallel.hpp>
+#include <type_traits>
+
+namespace Kokkos::Experimental::Impl {
+
+using OpenACCIterateLeft  = std::integral_constant<Iterate, Iterate::Left>;
+using OpenACCIterateRight = std::integral_constant<Iterate, Iterate::Right>;
```
Comment (dalg24) [inline core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp:57]: Not used

**Review by dalg24 — COMMENTED** [review]
Did you try enabling more tests?

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp (hunk 12 above)
Comment (seyonglee) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp]: Deleted.

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp (hunk 7 above)
Comment (seyonglee) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp]: Updated to take into account ReducerType

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp (hunk 6 above)
Comment (seyonglee) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp]: Refactored similar code patterns using macro functions.

**Comment by keitaTN** [comment]
The current packaging error is strange.  Is it our problem?

**Review by keitaTN — APPROVED** [review]

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp [hunk 13]
```diff
@@ -81,7 +83,8 @@ class Kokkos::Impl::ParallelReduce<Functor, Kokkos::RangePolicy<Traits...>,
     }
 
     ValueType val;
-    typename Analysis::Reducer final_reducer(&m_functor);
+    typename Analysis::Reducer final_reducer(
+        &ReducerConditional::select(m_functor, m_reducer));
```
Comment (dalg24) [inline core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp:87]: Done as part of #5680

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp (hunk 9 above)
Comment (dalg24) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp:110]: Moved to #5696

File: core/src/OpenACC/Kokkos_OpenACC_Team.hpp (hunk 11 above)
Comment (dalg24) [reply core/src/OpenACC/Kokkos_OpenACC_Team.hpp:84]: Same changes proposed in #5610
